### PR TITLE
Add debug dump scripts

### PR DESCRIPTION
**Pull request by rnystrom**
## Changes in this pull request

Got a little time on the plane to mess around with this idea. Very much want comments on this! Some questions:

- We should probably compile out parts of this using `#if DEBUG`.
  - Should I compile out all of the description methods?
  - Maybe I wrap the entire debug files w/ `#if DEBUG ... #endif` so none of it is loaded?
- @ryanolsonk I used [FLEX](https://github.com/Flipboard/FLEX/blob/master/Classes/Utility/FLEXHeapEnumerator.m) for searching the heap, lmk if you see any problems w/ this
  - Its lifted w/ only a few changes for my specific needs (including comments 😂)
  - Chalk us up for using [another bit](https://github.com/Instagram/IGListKit/blob/master/Source/Internal/IGListAdapterProxy.m#L77-L82) of your code
- Added a basic test so that the basics of this work and coverage doesn't tank
  - Anything else I should test?
- Is there more data we should be dumping?
- I have to track batch update state on the updater in order to dump it. If this seems dangerous lmk and I can revert that part
  - Though the batch update info is super useful
- Hit me w/ your nits on style and stuff

Here's an [example dump](https://gist.github.com/rnystrom/7a0f4d39bf9e1320953be1e99d0e0bf8).

Issue fixed: #310

### Checklist

- [x] All tests pass. Demo project builds and runs.
- [x] I added tests, an experiment, or detailed why my change isn't tested.
- [ ] ~~I added an entry to the `CHANGELOG.md` for any breaking changes, enhancements, or bug fixes.~~

### CODE REVIEW

**Comment by jessesquires** [comment]
🎉 

glanced through everything. looks rad.

**Comment by rnystrom** [comment]
Talked w/ @ryanolsonk who recommended I read through @kastiglione's code on https://github.com/facebook/chisel but I'm not able to find the method for searching the heap.

I _think_ this should be ok as-is, but still needs compiled out if not debugging (pls never ship this lol).

**Comment by rnystrom** [comment]
Alright updated w/ compile-out on the debug scripts. I think this is ready.

**Comment by kastiglione** [comment]
I haven't gotten around to pushing the heap search to the public chisel repo. That code is currently internal. I'm not sure what your future plans are, but as it is, you could also do this without heap scanning by having `IGListAdapter.m` track its live instances when built for debug.

**Comment by rnystrom** [comment]
@kastiglione that.... would probably be a lot simpler. Good call!

**Comment by rnystrom** [comment]
Importing for more eyeballs

File: Source/Internal/IGListDebugger.m [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+@implementation IGListDebugger
+
+static NSHashTable<IGListAdapter *> *livingAdaptersTable = nil;
+
++ (void)trackAdapter:(IGListAdapter *)adapter {
+#if IGLK_DEBUG_DESCRIPTION_ENABLED
+    if (livingAdaptersTable == nil) {
+        livingAdaptersTable = [NSHashTable weakObjectsHashTable];
+    }
+    [livingAdaptersTable addObject:adapter];
```
Comment (kastiglione) [inline Source/Internal/IGListDebugger.m]: Are `IGListAdapter` instances only created on main thread? For future consideration: an alternative to tracking using weak references (which involves a global lock and I believe may cause the objc runtime to use slow path in some cases), you could track `__unsafe_unretained` pointers and have them be removed them from the hash table from `-[IGListAdapter dealloc]`.

Comment (rnystrom) [inline Source/Internal/IGListDebugger.m]: Yup main thread affined. This is also a good idea though, I like it.

**Pull request closed without merging**